src/savegame.cpp

```diff
@@ -635,69 +643,1183 @@ void ingame_savegame::write_game(config_writer &out) {
+	ss << "        first_time_only=no\n";
+	ss << "        [modify_unit]\n";
+	ss << "            [filter]\n";
+	ss << "                side=$side_number\n";
+	ss << "            [/filter]\n";
+	ss << "            [object]\n";
+	ss << "            [effect]\n";
+	ss << "                apply_to=status\n";
+	ss << "                remove=fighting\n";
+	ss << "            [/effect]\n";
+	ss << "            [/object]\n";
+	ss << "        [/modify_unit]\n";
+	read(event, ss);
+}
+
+void healthy_trait_1_4_attack_end(config& event)
+{
+	std::stringstream ss;
+	ss << "        name=\"healthy attack end\"\n";
+	ss << "        name=\"attack end\"\n";
+	ss << "        first_time_only=no\n";
+	ss << "        [modify_unit]\n";
+	ss << "            [filter]\n";
+	ss << "                id=$unit.id\n";
+	ss << "            [/filter]\n";
+	ss << "            [object]\n";
+	ss << "                [effect]\n";
+	ss << "                    apply_to=status\n";
+	ss << "                    add=fighting\n";
+	ss << "                [/effect]\n";
+	ss << "            [/object]\n";
+	ss << "        [/modify_unit]\n";
+	ss << "        [modify_unit]\n";
+	ss << "            [filter]\n";
+	ss << "                id=$second_unit.id\n";
+	ss << "            [/filter]\n";
+	ss << "            [object]\n";
+	ss << "                [effect]\n";
+	ss << "                    apply_to=status\n";
+	ss << "                    add=fighting\n";
+	ss << "                [/effect]\n";
+	ss << "            [/object]\n";
+	ss << "        [/modify_unit]\n";
+	read(event, ss);
+}
+
+void update_prestart_detect_factions(config& cfg)
+{
+	if(config& replay = cfg.child("replay_start"))
+	{
+		for(config& event: replay.child_range("event"))
+		{
+			if(event["name"] == "prestart")
+			{
+				if(config& fire_event = event.child("fire_event"))
+				{
+					if(fire_event["name"] == "place_units")
+					{
+						config& lua = event.child("lua");
+
+						std::string changes[] = {"if searched == actual then;if actual:sub(1, #searched) == searched  then",
+												"if searched == actual.type then;if actual.type:sub(1, #searched) == searched then"};
+						std::ostringstream ss;
+				 		for (std::string& line : utils::split(lua["code"].str(), '\n'))
+				 		{
+							std::size_t found;
+							std::string oldcode, newcode;
+							for(auto it=std::begin(changes); it!=std::end(changes); ++it)
+							{
+								int sep = (*it).find(";");
+								oldcode = (*it).substr(0, sep);
+								newcode = (*it).substr(sep+1);
+								found = line.find(oldcode, 0);
+								if(found != std::string::npos)
+								{
+									line.replace(found, oldcode.length(), newcode);
+									break;
+								}
+							}
+							ss << line << "\n";
+						}
+						lua["code"] = ss.str();
+					}
+				}
+			}
+		}
+	}
+}
+
+void update_pre_recruited_loyals(config& cfg)
+{
+	const std::string suffix = convert_version_to_suffix(cfg);
+	if(suffix.empty()) return;
+	if(config& replay = cfg.child("replay_start"))
```

Comment:
I suspect many or most of these functions could be handled more cleanly with a WML patch file (of the sort created by `wesnoth --diff a b`). That'll obviously be quite a bit of work to create, but it would offload much of the burden of the update from the source code. It might even be more readable, though WML patch files aren't super readable either.